host.go

```diff
@@ -108,40 +125,129 @@ func (h *gpbftRunner) ValidateMessage(msg *gpbft.GMessage) (gpbft.ValidatedMessa
+	}
+	base.PowerTable, err = certs.MakePowerTableCID(pte)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("computing powertable CID for base: %w", err)
+	}
+
+	suffix := make([]gpbft.TipSet, min(gpbft.CHAIN_MAX_LEN-1, len(collectedChain))) // -1 because of base
+	for i := range suffix {
+		suffix[i].Key = collectedChain[i].Key()
+		suffix[i].Epoch = collectedChain[i].Epoch()
+
+		pte, err = h.client.ec.GetPowerTable(h.runningCtx, suffix[i].Key)
+		if err != nil {
+			return nil, nil, xerrors.Errorf("getting power table for suffix %d: %w", i, err)
+		}
+		suffix[i].PowerTable, err = certs.MakePowerTableCID(pte)
+		if err != nil {
+			return nil, nil, xerrors.Errorf("computing powertable CID for base: %w", err)
+		}
 	}
-	ptCid, err := certs.MakePowerTableCID(pt.Entries)
+	chain, err := gpbft.NewChain(base, suffix...)
 	if err != nil {
-		return nil, nil, xerrors.Errorf("computing power table CID: %w", err)
+		return nil, nil, xerrors.Errorf("making new chain: %w", err)
 	}
 
-	ts := sim.NewTipSetGenerator(1)
-	chain, err := gpbft.NewChain(
-		gpbft.TipSet{Epoch: 0, Key: ts.Sample(), PowerTable: ptCid},
-		gpbft.TipSet{Epoch: 1, Key: ts.Sample(), PowerTable: ptCid},
-	)
+	var supplData gpbft.SupplementalData
+	pt, _, err := h.GetCommitteeForInstance(instance + 1)
 	if err != nil {
-		return nil, nil, xerrors.Errorf("geenrating chain: %w", err)
+		return nil, nil, xerrors.Errorf("getting commite for %d: %w", instance+1, err)
 	}
-	sd := &gpbft.SupplementalData{
-		PowerTable: ptCid,
+
+	supplData.PowerTable, err = certs.MakePowerTableCID(pt.Entries)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("making power table cid for supplemental data: %w", err)
 	}
 
-	// TODO: use lookback to return the correct next power table commitment and commitments hash.
-	return sd, chain, nil
+	return &supplData, chain, nil
 }
 
 func (h *gpbftHost) GetCommitteeForInstance(instance uint64) (*gpbft.PowerTable, []byte, error) {
+	var powerTsk gpbft.TipSetKey
+	var powerEntries gpbft.PowerEntries
+	var err error
+
+	if instance < h.manifest.CommiteeLookback {
+		//boostrap phase
+		ts, err := h.client.ec.GetTipsetByEpoch(h.runningCtx, h.manifest.BootstrapEpoch-h.manifest.ECFinality)
+		if err != nil {
+			return nil, nil, xerrors.Errorf("getting tipset for boostrap epoch with lookback: %w", err)
+		}
+		powerTsk = ts.Key()
+		powerEntries, err = h.client.ec.GetPowerTable(h.runningCtx, powerTsk)
+		if err != nil {
+			return nil, nil, xerrors.Errorf("getting power table: %w", err)
+		}
+	} else {
+		cert, err := h.client.certstore.Get(h.runningCtx, instance-h.manifest.CommiteeLookback)
+		if err != nil {
+			return nil, nil, xerrors.Errorf("getting finality certificate: %w", err)
+		}
+		powerTsk = cert.ECChain.Head().Key
+
+		powerEntries, err = h.client.certstore.GetPowerTable(h.runningCtx, instance)
+		if err != nil {
+			// this fires every round, is this correct?
+			h.log.Infof("failed getting power table from certstore: %v, falling back to EC", err)
```

Comment:
We should never see an error here as long as we've already inserted the certificate for the previous instance. I wonder if we have an off-by-one somewhere? Or are we not inserting the cert?

Reply:
It is always off by one in the error message

Reply:
What's the specific message and/or can I reproduce it somehow?

Reply:
OH! It's from here:

https://github.com/filecoin-project/go-f3/blob/cd30c14975edb31dfeae832596d73579ab1cad0c/host.go#L314-L317

Reply:
This one: https://github.com/filecoin-project/go-f3/blob/main/certstore/certstore.go#L282

You can reproduce by running:
`./f3 manifest gen` and then `./f3 run --id=0` and `./f3 run --id=1` in two different terminals.